TDDWithURMAndKotlin/part2_increment.md

```diff
@@ -0,0 +1,157 @@
+
+Following the approach we followed for the `zero` function, we start by writing a test on how we expect the function behaves with negative numbers:
+
+```kotlin
+@ParameterizedTest
+@ValueSource(ints = [-2, -8, -20, -100])
+fun `increment function throws exception with negative position`(position: Int) {
+    assertThrows<IllegalArgumentException> { increment(registry, position) }
+
+    assertNull(registry.getValueAtPosition(position))
+}
+```
+
+This initial test is in pro of the defensive model, we first need to be sure the function fails for some invalid cases. We could start with a different part of the code but this is the easiest to write because we already wrote this on the `zero` function:
+
+```kotlin
+fun increment(registry: Registry, position: Int){
+    require(position > 0) { "Position must be positive number" }
+}
+```
+
+For our second test we need want to validate a "strict" rule: when a register has no value, it cannot be incremented. Our test should represent this and we should put a simple exception for it:
+
+```kotlin
+@ParameterizedTest
+@ValueSource(ints = [2, 4, 20, 40])
+fun `increment function cannot work with empty register`(position: Int) {
+    assertThrows<IllegalStateException> { increment(registry, position) }
+
+    assertNull(registry.getValueAtPosition(position))
+}
+```
+
+If we run our test at this point it will fail:
+
+```
+Expected java.lang.IllegalStateException to be thrown, but nothing was thrown.
+org.opentest4j.AssertionFailedError: Expected java.lang.IllegalStateException to be thrown, but nothing was thrown.
+```
+
+This exception tell us what we already know: our function is not throwing the right exception for empty registers. Let's fix that using the contract `checkNotNull` which is available on the kotlin stdlib (I'll write about contracts in another article):
+
+```kotlin
+fun increment(registry: Registry, position: Int){
+    require(position > 0) { "Position must be positive number" }
+    checkNotNull(registry.getValueAtPosition(position)) { "Register must be initialized" }
+}
+```
+
+Now our tests are passing again. Let's write the actual implementation.
+
+## The real implementation
+
+The test for the real implementation will need a litle more setup. We know that the current approach will throw an expection if our test is just:
+
+```kotlin
+@ParameterizedTest
+@ValueSource(ints = [11, 15, 110, 200, Int.MAX_VALUE])
+fun `increment function adds 1 to value of position`(position: Int) {
+    increment(registry, position)
+
+    assertEquals(1, registry.getValueAtPosition(position))
+}
+```
+
+This is because we wrote a test that will throw an exception when a registry does not contain a value and someone calls the `increment` function on it. But fortunately we already wrote tests for the `zero` function which means we can initialize our registry before calling increment:
```

Comment:
```suggestion
This is because we wrote a test that will throw an exception when a registry does not contain a value and someone calls the `increment` function on it. But fortunately we already wrote tests for the `zero` function which means we can initialize our registry before calling `increment`:
```